llvm/lib/Transforms/Scalar/SROA.cpp

```diff
@@ -5045,10 +5045,20 @@ static void insertNewDbgInst(DIBuilder &DIB, DbgVariableRecord *Orig,
                                                    BeforeInst->getIterator());
     return;
   }
+
+  if (Orig->isDbgValue()) {
+    DbgVariableRecord *DVR = DbgVariableRecord::createDbgVariableRecord(
+        NewAddr, Orig->getVariable(), NewFragmentExpr, Orig->getDebugLoc());
+    BeforeInst->getParent()->insertDbgRecordBefore(DVR,
+                                                   BeforeInst->getIterator());
```

Comment:
I was expecting some code checking for a DW_OP_deref and removing the DW_OP_deref here. Why is this not needed? Is it because we currently (incorrectly) implicitly treat dbg.value(alloca) as having an implicit op_deref?

Reply:
In this case, we're transferring the dbg_record from an alloca to an alloca, so there's no need to change the expression. I believe popping the DW_OP_deref is handled in mem2reg, which will be invoked by SROA later on.

Reply:
Thanks, that addresses one half of my concern. But, shouldn't we still verify that the dbg.value describes the value inside the alloca? If a dbg.value doesn't have a deref, it describes the *address* of the allcoa, and I don't think this is something we can handle here.

Reply:
(maybe we could handle it with DW_OP_implicit_pointer, but let's not overcomplicate things now)

Reply:
I see what you mean - I think such a check should _probably_ exist at the level above this, i.e. the caller should detect such cases and kill the debug value before getting to this point.

Reply:
You marked this as resolved, but did you add the check/assert?

Reply:
Sorry about that, it is resolved now, there is one issue however, @SLTozer mentioned that the check for the DW_OP_deref should be in the caller for this function, but this function has 2 overloads, i.e. 3 variations total. The functions can take a `DbgVariableRecord`, `DbgAssignIntrinsic`, or a `DbgDeclareInst`. We only care about the `DbgVariableRecord` case, but the caller doesn't distinguish between them because it is a lambda function that can take any of the three options as an input. So I had to handle the dropping of the debug info in the function itself rather than it's caller

Reply:
Just a note for the future based on a point @jmorse raised to me offline - if we have a dbg.value that pointed at the whole alloca being split, then it should be deleted here (but could be saved by implicit_pointer in theory), but if a dbg.value was pointing at one of the split fragments of the alloca, it can still have a valid address after this. I think that's something to implement in another patch rather than adding to this one, though.